### ISSUE TITLE
Traverse strategy breadth_first won't work with LIMIT keyword

### ISSUE THREAD

**Issue by giacomoParigi** [orientechnologies/orientdb#5330]
I already posted a question in the issue tracker here : https://github.com/orientechnologies/orientdb/issues/5314 , but I realized that the question is too wide and decided to try and break it into smaller issues.

The first issue is a bug on the parsing of the Traverse operator. If the keyword `<limit>` is added, both with -1 or with an actual positive limit, the statement `<strategy breadh_first>` is ignored in turn, as shown by the results of these two simple queries on the GratefulDeadConcerts default database (start node is chosen randomly):

First query (without `<limit>`):

```
  select $current.traversedVertex(-1), $path from
  (
  traverse * from #9:24
    while $depth < 4

    strategy breadth_first
  ) 
```

Results (only first 20):

$current | $path
------------|----------
#9:24 | (#9:24)
#9:15 | (#9:24).in_followed_by[0](#11:75)
#9:15 | (#9:24).in_followed_by[1](#11:76)
#9:15 | (#9:24).in_followed_by[2](#11:1476)
#9:15 | (#9:24).in_followed_by[3](#11:2356)
#9:15 | (#9:24).in_followed_by[4](#11:5414)
#9:15 | (#9:24).out_followed_by[0](#11:5416)
#9:15 | (#9:24).out_followed_by[1](#11:5420)
#9:15 | (#9:24).out_followed_by[2](#11:5424)
#9:15 | (#9:24).out_followed_by[3](#11:5425)
#9:15 | (#9:24).out_followed_by[4](#11:5431)
#9:15 | (#9:24).out_followed_by[5](#11:5436)
#9:15 | (#9:24).out_followed_by[6](#11:5445)
#9:15 | (#9:24).out_followed_by[7](#11:5454)
#9:15 | (#9:24).out_followed_by[8](#11:5478)
#9:15 | (#9:24).out_followed_by[9](#11:5481)
#9:15 | (#9:24).out_followed_by[10](#11:5501)
#9:15 | (#9:24).out_followed_by[11](#11:5510)
#9:15 | (#9:24).out_followed_by[12](#11:5531)
#9:15 | (#9:24).out_followed_by[13](#11:5543)

Second query (with `<limit>`):

```
  select $current.traversedVertex(-1), $path from
  (
  traverse * from #9:24
    while $depth < 4
	limit -1
    strategy breadth_first
  ) 
```

Results (only first 20):

$current | $path
------------|----------
#9:24 | (#9:24)
#9:15 | (#9:24).in_followed_by[0](#11:75)
#9:15 | (#9:24).in_followed_by[0](#11:75).out(#9:15)
#9:3 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[0](#11:20)
#9:4 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[1](#11:21)
#9:5 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[2](#11:22)
#9:11 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[3](#11:23)
#9:12 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[4](#11:24)
#9:14 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[5](#11:25)
#9:17 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[6](#11:294)
#9:16 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[7](#11:505)
#9:92 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[8](#11:869)
#9:109 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[9](#11:870)
#9:54 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[10](#11:871)
#9:83 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[11](#11:872)
#9:48 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[12](#11:873)
#9:53 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[13](#11:874)
#9:41 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[14](#11:875)
#9:160 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[15](#11:876)
#9:146 | (#9:24).in_followed_by[0](#11:75).out(#9:15).in_followed_by[16](#11:877)

**Comment by luigidellaquila** [comment]
Hi @giacomoParigi 

you should apply the LIMIT clause to the outer query, otherwise Studio will apply a 20 by default,
eg.

```sql
  select $current.traversedVertex(-1), $path from
  (
  traverse * from #9:24
    while $depth < 4
    strategy breadth_first
  ) 
  limit -1
```

**Comment by luigidellaquila** [comment]
I'm closing this, please reopen in case I missed something

Thanks

Luigi

**Comment by giacomoParigi** [comment]
The problem is not on the *number* of results (in that case you would be right pointing at the position of the ```LIMIT``` clause), but on their *order*. When the ```LIMIT``` clause is applied to the inner query - i.e. the ```TRAVERSE``` - the strategy falls back to the default ```depth_first```, ignoring the following ```strategy``` statement.

**Comment by giacomoParigi** [comment]
To replicate it should be enough to copy and paste the queries - the results are from the public GratefulDeadConcerts database - but if it is useful I can add a scripted toy example.

**Comment by luigidellaquila** [comment]
ouch... ok, reopening

**Comment by luigidellaquila** [comment]
Ciao @giacomoParigi 

I just pushed a fix on 2.1.x branch, it was a trivial parsing problem.
Now I'm porting the fix to develop branch

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Ok, it's also fixed in develop branch.
I'm closing this, please reopen in case something is missing

Thanks

Luigi
